src/meili/map_matcher.cc

```diff
@@ -532,45 +532,59 @@ void MapMatcher::RemoveRedundancies(const std::vector<StateId>& result)
 std::vector<MatchResults>
 MapMatcher::OfflineMatch(const std::vector<Measurement>& measurements, uint32_t k)
 {
+  if (k <= 0) {
+    throw std::invalid_argument("expect k to be positive but got " + std::to_string(k));
+  }
+
   // Reset everything
   Clear();
 
+  std::vector<MatchResults> best_paths;
+
   // Nothing to do
-  if (measurements.empty())
-    return {};
+  if (measurements.empty()) {
+    best_paths.emplace_back(
+        std::vector<MatchResult>{},
+        std::vector<EdgeSegment>{},
+        -1.0);
+    return best_paths;
+  }
+
+  bool found_broken_path = false;
 
   // Separate the measurements we are using for matching from the ones we'll just interpolate
   auto interpolated = AppendMeasurements(measurements);
 
-  // Get the time at the last column of states
-  auto time = container_.size() - 1;
-
   //For k paths
-  std::vector<MatchResults> best_paths;
-  std::vector<StateId> state_ids, original_state_ids;
-  while(best_paths.size() < k) {
-    // If we just got some results then we need to remove them from consideration
-    // we avoid doing this for the common case where k==1 by putting this at the start of the loop
-    if(best_paths.size()) {
-      // Remove all the candidates pairs whose paths are redundant with this one
-      RemoveRedundancies(original_state_ids);
-      // Remove this particular sequence of stateids
-      ts_.RemovePath(state_ids);
-      // Prepare for a fresh search in the next search iteration
-      vs_.ClearSearch();
-    }
-
+  while(best_paths.size() < k && !found_broken_path) {
     // Get the states for the kth best path in reversed order then fix the order
-    state_ids.clear();
-    std::copy(vs_.SearchPath(time, best_paths.empty()), vs_.PathEnd(), std::back_inserter(state_ids));
+    std::vector<StateId> state_ids;
+    double accumulated_cost = 0.f;
+    while (state_ids.size() < container_.size()) {
+      // Get the time at the last column of states
+      const auto time = container_.size() - state_ids.size() - 1;
+      std::copy(vs_.SearchPath(time, false), vs_.PathEnd(), std::back_inserter(state_ids));
+      const auto& winner = vs_.SearchWinner(time);
+      if (winner.IsValid()) {
+        accumulated_cost += vs_.AccumulatedCost(winner);
+      } else {
+        accumulated_cost += MAX_ACCUMULATED_COST;
+        found_broken_path = true;
+      }
 
-    // If we ended up finding a break in the path and this is not the first result we are done
-    if(state_ids.empty() || state_ids.back().time() != 0) {
-      break;
+      if (state_ids.size() < container_.size()) {
+        found_broken_path = true;
+        accumulated_cost += MAX_ACCUMULATED_COST;
+      }
     }
 
+    // TODO: do something to get real cost later, if we dont like sending back cost from here we can simply send
```

Comment:
Not sure what it means.

Reply:
its old. you can remove it. imho only real TODO left is to come up with something better than `MAX_ACCUMULATED_COST`